### PR TITLE
Re-enable X11 components on OS X

### PR DESCRIPTION
**Pull request by pkgw**
Now that we have the cross-platform X.org libraries available.

We'll see how the CI goes ...

### CODE REVIEW

File: recipe/build.sh [hunk 1]
```diff
@@ -1,28 +1,24 @@
...
-    --disable-gtk-doc \
-    $XWIN_ARGS
+configure_args=(
+    --prefix="$PREFIX"
+    --enable-warnings
+    --enable-ft
+    --enable-ps
+    --enable-pdf
+    --enable-svg
+    --disable-gtk-doc
```
Comment (jakirkham) [inline recipe/build.sh]: Why this change to the configure args?

Comment (pkgw) [inline recipe/build.sh]: It's just a better way to deal with them, in my opinion. You can extend the lists a little more cleanly if there are platform-specific changes to make ... which is what inspired me to edit them here, except the point of the commit was to remove the platform-specific changes!

The real benefit, though, is that it lets you have an editable list of arguments when some of them have internal spaces, like can easily happen with things along the lines of `--cflags="$CFLAGS"`. When you use `bash` lists you can extend the list of arguments *and* preserve the internal spaces correctly upon expansion with `"${list[@]}"`.

Comment (jakirkham) [inline recipe/build.sh]: I'm ambivalent, but have been less active here. If it helps you with your work here, then I'm +1.

**Comment by jakirkham** [comment]
Any thoughts @ccordoba12?

Also, even though he isn't a maintainer here, I'd be curious about @mingwandroid's thoughts generally on integration of X11 throughout the stack. Especially given we had discussed this in the conference call.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -32,6 +32,12 @@ requirements:
     - libpng >=1.6.28,<1.7
     - zlib 1.2.*
     - icu 58.*
+    - xorg-libx11 1.6.*
+    - xorg-libxext 1.3.*
+    - xorg-libxrender 0.9.*
+    - xorg-renderproto
+    - xorg-xextproto
+    - xorg-xproto
```
Comment (ccordoba12) [inline recipe/meta.yaml]: Is this really needed on Windows? Cairo is working just fine on that OS without relying on xorg libraries.

Comment (ccordoba12) [inline recipe/meta.yaml]: I mean, I can see the need of this change on macOS but not on Windows

**Comment by mingwandroid** [comment]
Enabling X11 without providing libs causes problems with headless servers which tend not to have X11 libs installed.

Providing our own builds of X11 could mitigate that but could lead to failure to open a display instead.

It's really tricky, I'd err on the side of caution and try not to link X11 to things that don't strictly need it.

Shouldn't we all be using Wayland by now? (j/k)

**Comment by mingwandroid** [comment]
For OSX I'd definitely not enable X11 on Cairo or anywhere where there's any alternative.

X11 on OSX is horrible. When I added Cairo support to R I took care too avoid X11 altogether.

**Comment by mingwandroid** [comment]
Finally this might be something for the variants stuff that @mcg1969 work on.

**Comment by pkgw** [comment]
@mingwandroid Do you have a more specific set of things to look out for on X11 on OSX? Since we're providing the client libraries, the experience with a well-engineered library like Cairo should be smooth: X11 will be an optional backend, with a very clear path to auto-detection: whether `$DISPLAY` is set or not. But I don't have a ton of experience in how these things work out in practice.

**Comment by mingwandroid** [comment]
I believe optional dependencies are either coming or here already for conda. @kalefranz will know better.

How are the libraries handled then? Does Cairo end up dynamically loading one or the other depending on existence and application preference? Where does it search for these then? Will it pick up incompatible system versions in some case? Regardless for Cairo I do not *think * there is a benefit to X11 support. You may be able to support some file formats that no one uses anymore, any thing else? I'm genuinely asking here as I don't know.

**Comment by pkgw** [comment]
With this PR, the cairo libraries just pull in the conda-forge provided X11 libraries as dependencies. I have a hunch that you're thinking that you need to link to *system-provided* X11 libraries for things to work, but that's not the case; X11 is a protocol, and one set of client libraries can speak it just as well as another. (In fact, the X.org ones are pretty much the only maintained versions so they're likely the unilaterally best choice on any OS.)

As for the benefit, there's a lot of code out there that uses Cairo as a generic backend to render graphics to either file formats like PDF, or interactive graphics like an X11 system. It is not uncommon for scientific packages to be a bit behind the times and require this on OSX even if they could in theory render via Quartz. My understanding is that this is precisely the situation with the [pyngl feedstock](https://github.com/conda-forge/pyngl-feedstock/pull/3), which is blocked on OSX because of the lack of Cairo+X11.

**Comment by kalefranz** [comment]
Optional dependencies are my personal #1 must have for conda 4.4.

**Comment by mingwandroid** [comment]
I wasn't under that impression @pkgw. Any time I've used an X11 program on macOS the experience has grated somewhat.

Cairo just generates bitmaps, more or less. It'll write them to various surface formats (X11 surfaces being one, raw memory surfaces being another).

pyngl does seem to only support creating X11 windows, but that should be fairly orthogonal to Cairo, except for possibly needing to write a function to convert from raw memory surfaces to the X11 surfaces that the pyngl windows will require.

Here I feel you are introducing very heavy X11 dependencies on an otherwise very lightweight 2d graphics library because one particular consumer of that graphics library has no support for Cocoa.

Many other consumers of Cairo would not like the X11 dependencies.

That's just my 2c anyway.

**Comment by ocefpaf** [comment]
> Here I feel you are introducing very heavy X11 dependencies on an otherwise very lightweight 2d graphics library because one particular consumer of that graphics library has no support for Cocoa.

There are more packages that will need this. From the top of my head I can remember `pyferret` and `ncl` (probably more).

> Many other consumers of Cairo would not like the X11 dependencies.

It would be nice to list them so we can make a decision. It is unclear to me how to conciliate both :confused: 
(I would like to avoid a `cairo-x11` packages but if that is the only way to go...)

**Comment by mingwandroid** [comment]
That's why I had hoped that Cairo would be dynamic about this. A lot of libraries are moving that way as dynamic loading gets around all sorts of legal and technical issues.

**Comment by pkgw** [comment]
When you have XRender support, as we do here, Cairo can do X11 rendering using vector graphics primitives, not just bitmaps.

It's not thrilling to use X11 on OSX, but lots of scientific software is in a place where it's the only available option. A smaller subset of that software is Cairo-based, but as outlined above, that subset is non-empty.

I don't think it's accurate to describe the dependencies as "heavyweight". You download and install a few more MB of libraries. Compared to the standard Anaconda package set, that's negligible. These libraries impose zero cost unless you happen to have software that requires Cairo+X11, in which case you go from being unable to use that software on OSX, to being able to use it.

**Comment by jakirkham** [comment]
So I'm a little concerned that the discussion on X11 is getting fragmented. I think we need to open an issue on the webpage repo add/or add a CFEP stating how we plan to add X11. This should help consolidate the discussion and help to make a uniform decision about how to integrate it into the stack. As it is currently, I'm worried we will end up with some partial adoption of X11, which IMHO is worse than having none at all.

**Comment by pkgw** [comment]
This does seem like a very good candidate for a CFEP.

Relatedly, there's no such thing as a Conda-Forge Packaging Guidelines document, right?

**Comment by mingwandroid** [comment]
SVG in Cairo is independent of the X11 surface backend:

```
$ otool -L /Users/rdonnelly/mc-x64-3.5/lib/libcairo.dylib
/Users/rdonnelly/mc-x64-3.5/lib/libcairo.dylib:
	@rpath/libcairo.2.dylib (compatibility version 11403.0.0, current version 11403.8.0)
	@rpath/libpixman-1.0.dylib (compatibility version 35.0.0, current version 35.0.0)
	@rpath/libfontconfig.1.dylib (compatibility version 11.0.0, current version 11.2.0)
	@rpath/libfreetype.6.dylib (compatibility version 18.0.0, current version 18.4.0)
	@rpath/libpng16.16.dylib (compatibility version 44.0.0, current version 44.0.0)
	@rpath/libz.1.dylib (compatibility version 1.0.0, current version 1.2.8)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 48.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 855.17.0)
```

```
$  nm ~/mc-x64-3.5/lib/libcairo.dylib  | grep svg
...
lots of stuff
...
```

If Qt on macOS ends up depending on X11 and running any significant amount of code through X11 it will be a huge inefficiency.

**Comment by jakirkham** [comment]
Definitely hear your concern about including X11 by default @mingwandroid. Though I also hear @pkgw's need for X11 and it would be an improvement in the case of Linux where we have the kludgy `yum_requirements.txt`. I think the challenge now is determining how we want to reconcile those two in the stack overall. So maybe we can pause the Cairo specific discussion until @pkgw has a chance to write something up.

**Comment by jakirkham** [comment]
Also if this helps @pkgw, feel free to start a [hackpad]( https://conda-forge.hackpad.com/ ) on the topic to iron out some ideas. Not required by any means, but a resource you should feel free to use if you need it.

**Comment by jakirkham** [comment]
> Relatedly, there's no such thing as a Conda-Forge Packaging Guidelines document, right?

Actually there are. One of the few pieces of documentation we have. 😄

ref: http://conda-forge.github.io/docs/guidelines.html

**Comment by pkgw** [comment]
@mingwandroid I'm confused by your comment. Clearly SVG is independent — X11 is a totally separate backend. Things that don't explicitly ask for it won't use it. Qt on OSX will be unchanged since it doesn't want to us X11. Some extra dependent shared libraries will be linked in but that's it.

**Comment by pkgw** [comment]
I have started [drafting a CFEP](https://github.com/pkgw/conda-forge-enhancement-proposals/blob/pr-xorg/cfep-04.md). I haven't started writing the meat of the document yet but it's non-empty.

**Comment by pkgw** [comment]
First-draft CFEP has been [turned into a pull request](https://github.com/conda-forge/conda-forge-enhancement-proposals/pull/7). Discussion can proceed there.

**Comment by pkgw** [comment]
I would really like to unblock this. What are the prospects for getting a chance to hash out CFEP 04?

**Comment by jakirkham** [comment]
As an FYI, this will block building GTK on macOS.

**Comment by mingwandroid** [comment]
X11 is so utterly ugly on macOS, please do not drag it into everything that uses Cairo. Cairo is a very low level rendering library that's going to become the 2d renderer for C++. This is just so wrong guys!

**Comment by mingwandroid** [comment]
And it does not 'hold up' anything. Make an cairo-x11 if you insist on his path.

**Comment by ocefpaf** [comment]
> X11 is so utterly ugly on macOS, please do not drag it into everything that uses Cairo. Cairo is a very low level rendering library that's going to become the 2d renderer for C++. This is just so wrong guys!

Indeed it is ugly but the scientific software I mentioned above and probably @pkg's too, are not really worry about that. However, that does not mean we should force it in the ecosystem. I really do not want that!
Like I said before I want to find a solution that can be used on these end point software and not at the base of our dependencies.


> This is not correct. Make a cairo-x11 if you insist on his path.

If that is a possibility let's try that. But like I said before this discussion is side-tracking the real issue and more non-related topics are getting dragged here. Let's focus on what can be done and not on what we are never going to do.

**Comment by pkgw** [comment]
@mingwandroid I've reread your comments on this topic many times and I feel like we are just not operating from the same premises. So in an attempt to make things explicit, here's my rundown of the relevant technical facts *as I believe them to be*. I've tried to state them as neutrally as I can.

1. Cairo is a 2D rendering library. Library users make function calls along the lines of "draw a rectangle" to create graphics.
2. Cairo supports various rendering backends. Depending on the "surface" object underlying the drawing context used when the "draw a rectangle" API call is made, the rectangle might be rasterized into a user-owned pixel buffer, expressed in an SVG or PDF file, or drawn to screen via a graphics system like X11 or Quartz.
3. Pieces of software that use Cairo actively choose which backend they want when they create their drawing context, by deciding which type of surface to use. [Here are the available surfaces](https://www.cairographics.org/manual/cairo-surfaces.html).
3. High-level GUI toolkits like Gtk and Qt use Cairo as their rendering system. They choose which Cairo backend to use based on the active OS, compile-time availability, etc.
4. By building the main `cairo` package with X11 support on OSX, we make the X11 backend available on OSX.
5. This injects dependencies on several X11 shared libraries into the main `libcairo`. On my Fedora system I count 6 X11-related `DT_NEEDED` entries in the ELF dynamic linking table.
6. This of course means that such a Conda package needs to pull in dependencies on the relevant X11 packages.
7. Adding the X11 backend allows us to provide several pieces of software for OSX that do not otherwise work.
8. Adding the X11 backend *may* add new features to other existing OSX packages if they have conditional support for the X11 backend that's currently turned off due to its absence. I'm not aware of any examples of this.
9. OK, this one is something that I cannot claim to be absolutely sure of, but I am extremely confident of: high-level GUI toolkits such as Gtk and Qt will completely ignore the X11 backend on OSX because it is not the native drawing system for that platform.
10. Likewise, I am extremely confident that if somehow a Conda package of Gtk or Qt on OSX does try to use the X11 backend, it will be because we have misconfigured the build somehow. E.g. [here is Gtk+ having a configuration option to disable X11](https://github.com/GNOME/gtk/blob/master/meson_options.txt#L3).

I have some subjective conclusions that I derive from this factual understanding but in the interests of trying to make sure that we're on the same page — do the above all sound correct to you? Are there technical facts that you think are relevant that I haven't included?

**Comment by mingwandroid** [comment]
I am asking the we do not further break binary compatibility with defaults. Your points are not entirely correct, if I were to install X11 into a prefix and build many bits of software, that software *will* detect the headers and libs and build a very non-native version of that software that doesn't integrate with the system well at all.

Please use variants/features for this or hold off on it until I have more time to do a proper analysis of the downsides of using a custom X11. Binary compatibility with defaults and bloated size are, in my opinion, good enough reasons to not link X11 to the default `cairo` or `tk`.

**Comment by ocefpaf** [comment]
> I am asking the we do not further break binary compatibility with defaults. 

Don't worry that is in our sights.

> Please use variants/features for this or hold off on it until I have more time to do a proper analysis of the downsides of using a custom X11

No need to ask to hold off. Again, nothing will be done before we assess the consequences correctly.

BTW I forgot about features, maybe that is the path forward, but again let's talk about that in the next meeting.

**Comment by pkgw** [comment]
BTW, as far as I can tell, conda-build's "split output packages" feature is *almost* perfectly suited to the goal of segregating X11 support into a separate `cairo-x11` package. That is, Cairo's support for X11 almost completely comes in terms of extra files that can be dropped on top of a base package, or not.

The only potential problem I see is that the main `cairo-features.h` header file does include a few #defines that touch on X11, e.g. `#define CAIRO_HAS_XCB_SURFACE 1`. Any package that wants X and cared about these defines would have a build+run-time dependency on `cairo-x11` so things would work out with minimal effort. I think it's quite unlikely that there are any packages that gate their X11 support solely on these defines, though; the standard check would be for the cairo-x11 pkg-config file, plus almost everyone has some kind of configuration flag to override the auto-detection result.

Anyway, what's the status of split packages in conda-forge? Blocking on conda-build 3.x, right?

**Comment by ocefpaf** [comment]
> Anyway, what's the status of split packages in conda-forge? Blocking on conda-build 3.x, right?

Not a blocker, we can by-pass the pinning in conda-build-setup on a case-by-case basis until we migrate entirely to `conda-build 3`.

**Comment by jakirkham** [comment]
Could you please outline what analysis you would like to do @mingwandroid? Maybe others can help chip in on some of these steps.

**Comment by jakirkham** [comment]
It looks like we are running into issues by not using our own X11 as noted in PR ( https://github.com/conda-forge/cairo-feedstock/pull/32 ). Also it seems like there wasn't a response to the analysis needed to determine using our own X11 was ok (at least not in this thread). Could I ask that we revisit this PR?

**Comment by pkgw** [comment]
I would like to get this resolved one way or another. But FWIW I was a bit sloppy in my referencing — if the Linux X11 linking issues are indeed best resolved by using our own X libraries, we can go that route without bringing them in on other platforms.

**Comment by pkgw** [comment]
This has been stale for a long time. I'm gonna go ahead and close so it leaves my todo queue. We can always reopen if needed.
